### PR TITLE
feat: cleanTempDir

### PR DESCRIPTION
**Pull request by kmdrGroch**
fixes: #2435

### CODE REVIEW

**Comment by kmdrGroch** [comment]
@nicojs something like this? 🤔

File: packages/core/src/utils/TemporaryDirectory.ts [hunk 1]
```diff
@@ -12,10 +12,12 @@ import { deleteDir } from './fileUtils';
 export class TemporaryDirectory implements Disposable {
   private readonly temporaryDirectory: string;
   private isInitialized = false;
+  private readonly removeDuringDisposal: boolean;
 
   public static readonly inject = tokens(commonTokens.logger, commonTokens.options);
   constructor(private readonly log: Logger, options: StrykerOptions) {
     this.temporaryDirectory = path.resolve(options.tempDirName);
+    this.removeDuringDisposal = options.cleanTempDir ?? true;
```
Comment (nicojs) [inline packages/core/src/utils/TemporaryDirectory.ts:20]: This `??` isn't needed right? Since it has a default value.

**Review by nicojs — CHANGES_REQUESTED** [review]
Seems good. One remark.

I've been thinking. Probably the best user experience is to leave the temp dir when an error occurred by default (our old behavior). For example, when `cleanTempDir` is true, only delete it when mutation testing was successful. 

To do this we would need to communicate the fact that an error occurred from `Stryker`'s `catch`. 

@kmdrGroch do you agree? If you need help, I could also implement this with a commit here.

**Comment by kmdrGroch** [comment]
Updated:
- now the functionality should be the same as it was before #2315 with a possibility to overwrite :)

**Comment by nicojs** [comment]
I think the old behavior is still not there,. The `dispose` method of the `TemporaryDirectory` is called no matter what (error or no error), see https://github.com/stryker-mutator/stryker/blob/6885e16fad7699ba93e6ebbbf0755c7d98c50c5a/packages/core/src/Stryker.ts#L59.

**Comment by kmdrGroch** [comment]
right! I am a dummy :D

**Comment by nicojs** [comment]
>  If you want removing all disposes, i think what should be done is adding "enableDisposal" option or sth like this 🤔

No, we need to `dispose`, otherwise Stryker hangs forever 🙄. `dispose()` makes sure all child processes are closed and nodejs is allowed to exit.

What I meant was something like this:

Inside Stryker.ts

```ts
// 1. Prepare. Load Stryker configuration, load the input files and starts the logging server
      const prepareExecutor = loggerProvider.provideValue(coreTokens.cliOptions, this.cliOptions).injectClass(PrepareExecutor);
      const mutantInstrumenterInjector = await prepareExecutor.execute();

      try {
        // 2. Mutate and instrument the files and write to the sandbox.
        const mutantInstrumenter = mutantInstrumenterInjector.injectClass(MutantInstrumenterExecutor);
        const dryRunExecutorInjector = await mutantInstrumenter.execute();

        // 3. Perform a 'dry run' (initial test run). Runs the tests without active mutants and collects coverage.
        const dryRunExecutor = dryRunExecutorInjector.injectClass(DryRunExecutor);
        const mutationRunExecutorInjector = await dryRunExecutor.execute();

        // 4. Actual mutation testing. Will check every mutant and if valid run it in an available test runner.
        const mutationRunExecutor = mutationRunExecutorInjector.injectClass(MutationTestExecutor);
        const mutantResults = await mutationRunExecutor.execute();

        return mutantResults;
      } catch (error) {
       // Mark current run as failed, so tempoary directory can choose if it needs to clean or not.
        mutantInstrumenterInjector.resolve(coreTokens.temporaryDirectory).markRunErrored();
        throw error; // rethrow
      }
```

**Comment by nicojs** [comment]
Is this clear @kmdrGroch ? I don't see another way, other than adding a parameter to the `dispose` and alter `typed-inject` to accomondate this. Not sure yet if I would want to add that.

**Comment by kmdrGroch** [comment]
I don't quite get it 😕

**Comment by nicojs** [comment]
I had some time, so implemented it.

What I did:

* Made sure to only clean the temp dir when Stryker succesfully executed (it's quite simple, please have a look 👍)
    ```ts
    catch (error) {
      const log = loggerProvider.resolve(commonTokens.getLogger)(Stryker.name);
      log.debug('Not removing the temp dir because an error occurred');
      mutantInstrumenterInjector.resolve(coreTokens.temporaryDirectory).removeDuringDisposal = false;
      throw error;
    }      
    ```
* Made `true` the default value (so after a succesful run, the `.stryker-tmp` dir is gone.
* I've added `--cleanTempDir` as a cli option
* I've documented it in the readme.md

**Review by nicojs — APPROVED** [review]

**Pull request merged**